finisher_api.go

```diff
@@ -181,6 +181,21 @@ func (db *DB) FindInBatches(dest interface{}, batchSize int, fc func(tx *DB, bat
 		batch        int
 	)
 
+	// user specified offset or limit
+	var totalSize int
+	if c, ok := tx.Statement.Clauses["LIMIT"]; ok {
+		if limit, ok := c.Expression.(clause.Limit); ok {
+			totalSize = limit.Limit
+
+			if batchSize > totalSize {
+				batchSize = totalSize
+			}
+
+			// reset to offset to 0 in next batch
+			tx = tx.Offset(-1).Session(&Session{})
```

Comment:
Should we extract this call to another check? The `offset` is also buggy when using w/o `limit`.

Reply:
> Should we extract this call to another check? The `offset` is also buggy when using w/o `limit`.

What does `offset is also buggy` mean ? `FindInBatches` query condition is greater than primary key and limit.

Reply:
> > Should we extract this call to another check? The `offset` is also buggy when using w/o `limit`.
> 
> What does `offset is also buggy` mean ? `FindInBatches` query condition is greater than primary key and limit.

If we only using `offset`, which scenario will be like *select all users which uid is larger than 100*, then using `dbInstance.Offset(100).OrderBy("user_id").FindInBatches()`, it will skip 100 records in each of all subsequent queries.

Reply:
it will skip 100 records in first subsequent query.
here is reset offset to zero after first subsequent query that we can use `id > ? and limit = ?` to query.

Reply:
> it will skip 100 records in first sub query. here is reset offset to zero after first sub query that we can use `id > ? and limit = ?` to query.

That's correct. But this call just been embedded in the `limit` check. I mean, we may need to extract it to another check that if we have a `Offset` clause and the argument is large than `0` so we can both support only using `Offset` or togethered with `Limit` clause.

Reply:
`Offest` and `Limit` use the same `clause.Limit`, so it support for only using `Offset` or togethered
https://github.com/go-gorm/gorm/blob/master/chainable_api.go#L245

Reply:
Only `Offset` cannot be used is caused by [here](https://github.com/go-gorm/gorm/pull/5255/commits/f9a86e9148fab7af0f932d8c1350683087d794a4#diff-aaa805d49ddced2f2281789039fef3bdb2fca44a51914d2c9b05ea3bd959776aR190), thank you.

Reply:
> Only `Offset` cannot be used is caused by [here](https://github.com/go-gorm/gorm/pull/5255/commits/f9a86e9148fab7af0f932d8c1350683087d794a4#diff-aaa805d49ddced2f2281789039fef3bdb2fca44a51914d2c9b05ea3bd959776aR190), thank you.

Got it. You're nice :).